File: AULA 7/C#/C - The dominoes solitaire.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Text;
using System.Numerics;

public class Program
{
    public static void Main(string[] args)
    {

        while (true)
        {
            string input = Console.ReadLine();
            if (!int.TryParse(input, out var n) || n == 0)
                break;

            var m = int.Parse(Console.ReadLine());

            var pieces = new List<Piece>();

            for (int i = 0; i < m + 2; i++)
            {
                string[] values = Console.ReadLine().Split(' ');
                pieces.Add(new Piece (int.Parse(values[0]), int.Parse(values[1]), false));
            }

            var flag = new Flag(false);
            Dfs(pieces, n, m, 0, pieces[0].q, flag);

            Console.WriteLine(flag.value ? "YES" : "NO");
        }
    }

    public static void Dfs(List<Piece> pieces, int n, int m, int idx, int pieceNum, Flag flag)
    {
        if (idx == n)
        {
            if (pieceNum == pieces[1].p)
                flag.value = true;

            return;
        }

        for (int i = 2; i < m + 2; i++)
        {
            if (!pieces[i].isUsed)
            {
                if (pieceNum == pieces[i].p)
                {
                    pieces[i].isUsed = true;
                    Dfs(pieces, n, m, idx + 1, pieces[i].q, flag);
                    pieces[i].isUsed = false;
                }

                if (pieceNum == pieces[i].q)
                {
                    pieces[i].isUsed = true;
                    Dfs(pieces, n, m, idx + 1, pieces[i].p, flag);
                    pieces[i].isUsed = false;
                }
            }
        }
    }

    public class Piece(int p, int q, bool isUsed)
    {
        public int p { get; set; } = p;
        public int q { get; set; } = q;
        public bool isUsed { get; set; } = isUsed;
    }

    public class Flag(bool value)
    {
        public bool value { get; set; } = value;
    }
}

File: AULA 7/Java/C - The dominoes solitaire.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        while (true) {
            int n = scanner.nextInt();

            if (n == 0) {
                break;
            }

            int m = scanner.nextInt();

            List<Piece> pieces = new ArrayList<>();

            for (int i = 0; i < m + 2; i++) {
                int p = scanner.nextInt();
                int q = scanner.nextInt();
                pieces.add(new Piece(p, q, false));
            }

            Flag flag = new Flag(false);
            dfs(pieces, n, m, 0, pieces.get(0).q, flag);

            System.out.println(flag.value ? "YES" : "NO");
        }

        scanner.close();
    }

    public static void dfs(List<Piece> pieces, int n, int m, int idx, int pieceNum, Flag flag) {
        if (idx == n) {
            if (pieceNum == pieces.get(1).p) {
                flag.value = true;
            }
            return;
        }

        for (int i = 2; i < m + 2; i++) {
            if (!pieces.get(i).isUsed) {
                if (pieceNum == pieces.get(i).p) {
                    pieces.get(i).isUsed = true;
                    dfs(pieces, n, m, idx + 1, pieces.get(i).q, flag);
                    pieces.get(i).isUsed = false;
                }

                if (pieceNum == pieces.get(i).q) {
                    pieces.get(i).isUsed = true;
                    dfs(pieces, n, m, idx + 1, pieces.get(i).p, flag);
                    pieces.get(i).isUsed = false;
                }
            }
        }
    }

    static class Piece {
        int p, q;
        boolean isUsed;

        public Piece(int p, int q, boolean isUsed) {
            this.p = p;
            this.q = q;
            this.isUsed = isUsed;
        }
    }

    static class Flag {
        boolean value;

        public Flag(boolean value) {
            this.value = value;
        }
    }
}
